zuul-core/src/main/java/com/netflix/zuul/message/Headers.java

```diff
@@ -195,6 +195,30 @@ public void setAndValidate(String headerName, @Nullable String value) {
         setNormal(validateField(headerName), validateField(normalName), validateField(value));
     }
 
+    /**
+     * Replace any/all entries with this key, with this single entry if the key and entry are valid.
+     *
+     * If value is {@code null}, then not added, but any existing header of same name is removed.
+     */
+    public void setIfValid(HeaderName headerName, String value) {
+        String normalName = requireNonNull(headerName, "headerName").getNormalised();
+        if (isValid(headerName.getName()) && isValid(normalName) && isValid(value)) {
```

Comment:
I think you should check if it is valid first, and then normalize.   I think it is true that if the non-normal name is valid, the normal one must be as well.  Thus, I dont think you need to validate the normal name.